pkg/kubelet/prober/runner.go

```diff
@@ -0,0 +1,335 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package prober
+
+import (
+	"context"
+	"fmt"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+	podutil "k8s.io/kubernetes/pkg/api/v1/pod"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	"k8s.io/kubernetes/pkg/probe"
+	httpprobe "k8s.io/kubernetes/pkg/probe/http"
+	"net/http"
+	"time"
+)
+
+// unknownProbeTypeError is used by dummyProbeRunner for returning error.
+var unknownProbeTypeError = fmt.Errorf("unknown probe type")
+
+// unknownProbeError is used by dummyProbeRunner for returning error.
+var unknownProbeHandlerError = fmt.Errorf("missing probe handler")
+
+// probeRunner is an interface for running probes, objects implementing
+// probeRunner encapsulates the logic of running probes.
+type probeRunner interface {
+
+	// sync ensures objects required for running probe are in sync, with probe
+	// specification and latest pod status (podIP can mutate for running pod).
+	sync(v1.Container, v1.PodStatus, probeType) error
+
+	// run holds the logic of running probe, run takes prober as a dependency
+	// to run the probe on http.Prober, tcp.Prober, grpc.Prober or exec.Prober.
+	run(context.Context, v1.Container, v1.PodStatus, probeType, *prober) (probe.Result, string, error)
+}
+
+// tcpProbeRunner holds state and logic for running TCP probes.
+type tcpProbeRunner struct {
```

Comment:
I would implement each runner in its own file to ease code comparison between them, and also separate from the interface.

Reply:
Will update.